File: src/Dibbidut/Classes/Ship.java
package Dibbidut.Classes;

import Dibbidut.Classes.Geometry.HPoint;
import Dibbidut.Classes.Handlers.AISToShipHandler;
import Dibbidut.Classes.Handlers.UpdateShipHandler;
import Dibbidut.Classes.InputManagement.AISData;
import Dibbidut.Interfaces.IShipDataHandler;
import Dibbidut.Interfaces.IDomain;
import math.geom2d.Vector2D;

import java.awt.*;
import java.util.Hashtable;

public class Ship extends Obstacle {
    public IDomain domain;
    public int mmsi;
    public int length;
    public int width;
    public int heading;
    public double longitude;
    public double latitude;
    public double cog;
    public double sog;
    public double manoeuvrability;
    public Shape conflictRegion;
    public Hashtable<String, String> warnings;

    private AISData currentData;

    public Ship(int mmsi) {
        super(new HPoint(0,0), new HPoint(0,0));

        this.mmsi = mmsi;
    }

    public Ship(HPoint position, HPoint velocity, Shape conflictRegion) {
        super(position, velocity);
        this.position = position;
        this.velocity = velocity;
        this.conflictRegion = conflictRegion;
    }

    public Ship(HPoint position, int length, int width, int heading) {
        super(position, new HPoint(0,0));
        this.position = position;
        this.length = length;
        this.width = width;
        this.heading = heading;

        domain = new ShipDomain(length, width, "Pentagon");
        domain.Update(sog, heading, position.getY(), position.getX());
    }

    public Ship(AISData data) {
        super(new HPoint(0,0), new HPoint(0,0));

        warnings = new Hashtable<>();
        currentData = data;

        IShipDataHandler handler = new AISToShipHandler(this, currentData, warnings);

        handler.Run();

        domain = new ShipDomain(length, width, "Pentagon");
        domain.Update(sog, heading, position.getY(), position.getX());
    }

    /**
     * Updates the given ship with new data
     * @param data The new data that the ship will be updated with
     */
    public void Update(AISData data) {

        AISData oldData = currentData;
        currentData = data;

        IShipDataHandler handler = new UpdateShipHandler(this, currentData, oldData, warnings);

        handler.Run();

        this.velocity.scale(10);

        domain.Update(sog, heading, position.getY(), position.getX());
    }

    @Override
    public boolean equals(Object obj) {
        if (obj.getClass() == Ship.class) {
            return this.mmsi == ((Ship) obj).mmsi;
        }
        else if (obj.getClass() == AISData.class) {
            return this.mmsi == ((AISData) obj).mmsi;
        }
        else {
            return super.equals(obj);
        }
    }
}

File: DataGenerator/classes/ship.py
import math


class Ship:
    MMSI = 0
    Heading = 0.0
    COG = 0.0
    SOG = 0.0
    Length = 0.0
    Width = 0.0
    UnitVector = []

    def __init__(self, MMSI, heading, x, y, COG, SOG, length, width):
        self.MMSI = MMSI
        self.Heading = heading
        self.X = x
        self.Y = y
        self.COG = COG
        self.SOG = SOG
        self.Length = length
        self.Width = width

    def get_unit_vector(self):
        if len(self.UnitVector) == 0:
            self.unit_vector()
        return self.UnitVector

    def unit_vector(self):
        degrees_as_radians = math.radians(-self.COG)
        out_x = math.cos(degrees_as_radians) * 0 - math.sin(degrees_as_radians) * 1
        out_y = math.sin(degrees_as_radians) * 0 + math.cos(degrees_as_radians) * 1
        self.UnitVector = [out_x, out_y]
